python/src/aac/execute/command_line.py

```diff
@@ -0,0 +1,190 @@
+def cli():
+    """The Architecture-as-Code (AaC) command line tool."""
+    pass
+
+
+@cli.result_callback()
+def output_result(result: ExecutionResult):
+    """Output the result of the command."""
+    error_occurred = not result.is_success()
+    secho(result.get_messages_as_string(), err=error_occurred, color=True)
+
+    if error_occurred:
+        sys.exit(result.status_code.value)
+
+
+def to_click_type(type_name: str) -> ParamType:
+    """Convert the named type to a type recognized by Click."""
+    if type_name == "file":
+        return Path(file_okay=True)
+    elif type_name == "directory":
+        return Path(dir_okay=True)
+
+    return types.__dict__.get(type_name.upper(), UNPROCESSED)
+
+
+def to_click_parameter(argument: AacCommandArgument) -> Parameter:
+    """Convert an AacCommandArgument to a Click Parameter."""
+    names = [argument.name] if isinstance(argument.name, str) else argument.name
+    args = dict(
+        type=to_click_type(argument.data_type), nargs=1, default=argument.default
+    )
+    return (
+        Option(
+            names,
+            help=argument.description,
+            show_default=True,
+            is_flag=argument.data_type == "bool",
+            **args,
+        )
+        if argument.name[0].startswith("-")
+        else Argument(names, **args)
+    )
+
+
+def handle_exceptions(plugin_name: str, func: Callable) -> Callable:  # noqa: C901
+    """Decorator to catch and handle exceptions in a function."""
+
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except LanguageError as e:
+            usr_msg = f"{e.message}{linesep}{e.location}"
+            return ExecutionResult(
+                plugin_name,
+                "exception",
+                ExecutionStatus.GENERAL_FAILURE,
+                [ExecutionMessage(usr_msg, MessageLevel.ERROR, None, None)],
+            )
+        except OperationCancelled as e:
+            return ExecutionResult(
+                plugin_name,
+                "exception",
+                ExecutionStatus.OPERATION_CANCELLED,
+                [ExecutionMessage(str(e), MessageLevel.ERROR, None, None)],
+            )
+        except ParserError as e:
+            usr_msg = f"The AaC file '{e.source}' could not be parsed.{linesep}"
+            if e.errors:
+                usr_msg = f"{usr_msg}The following errors were encountered:{linesep}"
+                for err in e.errors:
+                    usr_msg += f"  - {err}{linesep}"
+            if e.yaml_error:
+                usr_msg += f"The following YAML errors were encountered:{linesep}"
+                exc = e.yaml_error
+                if hasattr(exc, "problem_mark"):
+                    if exc.context is not None:
+                        usr_msg += f"  parser says{linesep} {str(exc.problem_mark)} {linesep}{str(exc.problem)} {str(exc.context)}{linesep}Please correct data and retry."
```

Comment:
Knew I made a comment about drilling into the location somewhere!! Since I was working in plugins today was seeing a lot of `ExecutionResult` so had the class pulled back up to see what the innards looked like again. Saw where the `get_messages_as_string()` had a drill down on the location so think the very last portion of my above comment can be ignored.